### ISSUE TITLE
[no-extraneous-class] new option to specifically allow abstract classes

### ISSUE THREAD

**Issue by 0xCLARITY** [typescript-eslint/typescript-eslint#2015]
We use classes that only hold static members as namespaces in our codebase. It would be great to have a rule that required the `abstract` keyword on those classes, so they could not be called with the `new` keyword and would not have a constructor function.

This rule would pair well with the [`no-extraneous-class`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-extraneous-class.md) rule, with `allowStaticOnly` set to true.

#### Example

Given a class like this:

```ts
class Utils {
  public static isValidString(arg: unknown): arg is string {
    return typeof arg === 'string'
  }
}
```

Report a violation, and allow auto-fixing to this:
```ts
abstract class Utils {
  public static isValidString(arg: unknown): arg is string {
    return typeof arg === 'string'
  }
}
```

**Comment by bradzacher** [comment]
> We use classes that only hold static members as namespaces in our codebase.

OOC, why not just use `namespaces`, or plain objects?

----

This could probably just be a new option on `no-extraneous-class`, like `allowStaticOnAbstractClassOnly`.

**Comment by hbergren** [comment]
> OOC, why not just use namespaces, or plain objects?

I'm not super familiar with namespaces, and maybe that's a better pattern for this use case. Right now, we use the [`no-namespace`](https://github.com/typescript-eslint/typescript-eslint/blob/v2.31.0/packages/eslint-plugin/docs/rules/no-namespace.md) rule, so I'm not sure how we would use namespaces here, and that rule made me feel like namespaces might be the wrong way to organize the code here.

We've also talked about just using objects, but we use the [`naming-convention`] rule to enforce `camelCase` for variables, and we want our static abstract class / namespaces to be in `PascalCase`. (Fairly weak argument I know haha).

> This could probably just be a new option on no-extraneous-class, like allowStaticOnAbstractClassOnly.

Definitely! That makes sense to me.

**Comment by bradzacher** [comment]
Namespaces look almost the same as classes, but have the added bonus of being a full block scope, so you're not just limited to only methods/properties - you can also export enums, types, etc.

```ts
namespace Utils {
    export type ValidString = string & { __brand: never };
    export function isValidString(arg: unknown): arg is ValidString {
        return typeof arg === 'string'
    }
}

const x: Utils.ValidString = Utils.isValidString('');
```

Though TBH I would question why you even need a class or a namespace at all.
If you just use modules, you get namespacing for free with the added bonus of being able to selectively import things if you want.
Compared to `namespace`s, these support more export styles (can do `export * from 'otherModule'`, for example) and have a nicer compiled output.

```ts
// Utils.ts
export function isValidString(arg: unknown): arg is string {
    return typeof arg === 'string'
}

// other file 1
import * as Utils from './Utils';

Utils.isValidString('');

// other file 2
import { isValidString } from './Utils';

isValidString('');
```

**Comment by bbarry** [comment]
```
import * as Utils from './Utils';
```

Would it be possible to annotate files as to how we expect them to be imported within a codebase?

for example something like

```
// @typescript-eslint: import-star Utils
```

at the top of this Utils.ts file?

**Comment by bradzacher** [comment]
there is not.
OOC why does it matter if someone does a namespace import or a named import?
I guess one could argue for consistency's case, but is there any other reason?

It would probably a candidate for a rule to add to [`eslint-plugin-import`](https://github.com/benmosher/eslint-plugin-import)

**Comment by bbarry** [comment]
hmm, found this: https://github.com/benmosher/eslint-plugin-import/issues/1547

**Comment by Peeja** [comment]
> OOC, why not just use namespaces, or plain objects?

In this case, I agree. I'm hitting a case, though, where I want classes specifically. I'm writing a kind of parser, and I want to allow clients to override handlers for different parts of the AST. So the "class" is a bundle of implementations, and I want client code to be able to override any of those implementations, while still being able to easily call the others, or even the original implementation. This doesn't *have* to be done with a static-only class: I'd rather do it with a plain object, but making the inheritance and the type checking work right takes a lot of complexity. It could also be a class that gets instantiated, but I specifically intend to have no state in the parser, so I'd rather have it not be an instance.

I'm still not totally certain that using an abstract, static-only class is the best path, but so far it seems like the best choice available. I'm currently using `"allowStaticOnly": true`, but I think it might be better to base it on `abstract`—if I wrote a non-abstract class with only static members, it would almost certainly be a mistake.

**Comment by Peeja** [comment]
Hmm. You know what, I take that all back. I really thought I had it working with static-only classes, but there are some oddities I didn't expect, and it turns out it's worth the complexity to use plain objects and define my own inheritance:

```ts
const inherit = <T extends object, O extends Partial<T>>(
  orig: T,
  overrides: O
) =>
  Object.assign(Object.create(orig) as T, overrides);
```

So: ignore my argument above. 😛

**Comment by JoshuaKGoldberg** [comment]
This has been open for over 4 years with very little user activity and only one 👍. It's for a style that very fewer TypeScript developers generally use. I'm going to go ahead and close it as wontfix.

If someone has a strong argument in favor of this feature addition, please do post it - we can always reopen / file a new issue.

Cheers!
